src/utils/MSFraggerAdapter.cpp

```diff
@@ -0,0 +1,859 @@
+    setValidFormats_(TOPPMSFraggerAdapter::param_out, ListUtils::create<String>("pep.xml,pepXML,tsv"), true);
+
+    // Path to database to search
+    registerInputFile_(TOPPMSFraggerAdapter::param_database, "<path_to_fasta>", "", "FASTA database file", true, false);
+    setValidFormats_(TOPPMSFraggerAdapter::param_database, ListUtils::create<String>("FASTA,fasta,fa,fas"), false);
+
+    // TOPP tolerance
+    registerTOPPSubsection_("tolerance", "Search Tolerances");
+
+    // Precursor mass tolerance and unit
+    _registerNonNegativeDouble(TOPPMSFraggerAdapter::param_precursor_mass_tolerance, "<precursor_mass_tolerance>", 20.0, "Precursor mass tolerance", false, false);
+    registerStringOption_(TOPPMSFraggerAdapter::param_precursor_mass_unit, "<precursor_mass_unit>", "ppm", "Unit of precursor mass tolerance", false, false);
+    setValidStrings_(TOPPMSFraggerAdapter::param_precursor_mass_unit, validUnits);
+
+    // Precursor true tolerance
+    _registerNonNegativeDouble(TOPPMSFraggerAdapter::param_precursor_true_tolerance, "<precursor_true_tolerance>", 0.0, "Precursor true tolerance. It is strongly recommended to set this feature!", false, false);
+    registerStringOption_(TOPPMSFraggerAdapter::param_precursor_true_unit, "<precursor_true_unit>", "ppm", "Unit of precursor true tolerance", false, false);
+    setValidStrings_(TOPPMSFraggerAdapter::param_precursor_true_unit, validUnits);
+
+    // Fragment mass tolerance
+    _registerNonNegativeDouble(TOPPMSFraggerAdapter::param_fragment_mass_tolerance, "<fragment_mass_tolerance>", 20.0, "Fragment mass tolerance", false, false);
+    registerStringOption_(TOPPMSFraggerAdapter::param_fragment_mass_unit, "<fragment_mass_unit>", "ppm", "Fragment mass unit", false, false);
+    setValidStrings_(TOPPMSFraggerAdapter::param_fragment_mass_unit, validUnits);
+
+    // Isotope error
+    registerStringOption_(TOPPMSFraggerAdapter::param_isotope_error, "<isotope_error>", "0", "Isotope correction for MS/MS events triggered on isotopic peaks. Should be set to 0 for open search or 0/1/2 for correction of narrow window searches", false, false);
+    setValidStrings_(TOPPMSFraggerAdapter::param_isotope_error, isotope_error_and_enzyme_termini);
+
+    // TOPP digest
+    registerTOPPSubsection_("digest", "In-Silico Digestion Parameters");
+
+    // Enzyme
+    StringList enzyme_names;
+    EnzymesDB::getInstance()->getAllNames(enzyme_names);
+    registerStringOption_(TOPPMSFraggerAdapter::param_search_enzyme_name, "<search_enzyme_name>", "Trypsin", "Name of the enzyme to be written to the pepXML file", false, false);
+    setValidStrings_(TOPPMSFraggerAdapter::param_search_enzyme_name, enzyme_names);
+
+    // Cut after
+    registerStringOption_(TOPPMSFraggerAdapter::param_search_enzyme_cutafter, "<search_enzyme_cutafter>", "KR", "Residues after which the enzyme cuts", false , false);
```

Comment:
you could use valid strings for this option and add the MSFragger Enzyme codes into share/Chemistry/Enzymes.xml

I think XTandem Adapter uses a getAllXTandemNames function

Reply:
Or otherwise you could add the most important ones to the description, how to specify no enzyme or unspecific cleavage for example

Reply:
There doesn't seem to be a MSFragger Enzyme code. One only specifies the cleavage site.
I could probably use the XTandemID to calculate the respective parameters for MSFragger:
Advantage: Standardized enzymes
Disadvantages: Error prone and artificially reduces the power of the MSFragger parameters.

Reply:
If one does not want cut, then the string with all amino acid residues can be specified for `enzyme_butnotbefore`. Unspecific cleavages does not seem to be possible.

Reply:
Hmm we could do 2 parameters.
Predefined enzyme: shows a selection of our enzymes in the DB. You can add cutting rules instead of an ID for MSFragger. Plus a selection "User-specified rule".
User-specified rule: Only used, when User-specified rule is selected.

Or omit the user-specified rule selection and override the standard enzyme, when the second param is given.

Reply:
unspecified cleavage might be able to be achieved by just listing all 20 AAs under "cutafter" and setting missed cleavages to max peptide length (-1?).